Introduce RichStore

moreal opened the pull request:
In this PR, it introduces `RichStore`, `IStore` implementation extended `DefaultStore`.
It works like `DefaultStore`, but it stores metadata like below, to search fast on `PutBlock`, `PutTransaction`.

- Transaction → Block
- Signer → Transaction
- UpdatedAddress → Transaction

By them, you can see the block a transaction is included through `blockRef` field.

![image](https://user-images.githubusercontent.com/26626194/73446263-c9a33980-439f-11ea-8405-59b8da37c92f.png)

Also listing transactions will be fast because it skips empty blocks and iterate only transactions directly. In my laptop environment, it takes just half. Of course, it can be slow in this structure if there was no empty block.

PS. it became to throw exceptions from GraphQL logic. 7f6ecf3

moreal committed:
Implement RichStore

moreal committed:
Use RichStore instead DefaultStore

moreal committed:
Use (BlockChain<T>, IStore) as graphql context

moreal committed:
Implement BlockRef in TransactionType

moreal committed:
Improve RichStore

- Seperate tasks between PutBlock and PutTransaction.
- Make some iterations pageable.

moreal committed:
Improve performance transactions with RichStore

When using RichStore, it improves the performance to transactions
which were filtered with signer, involved.

moreal committed:
Throw exceptions from graphql inner logic.

moreal committed:
Improve RichStore

- Improve performance to list transactions without signer, involved with RichStore
- Fix a bug where the order doesn't work because it doesn't ensure index

moreal committed:
Fix a bug where not sorted.

moreal committed:
Use RichStore in only online-mode

moreal committed:
Return IEnumerable directly

dahlia commented on Libplanet.Explorer/Store/AddressRefDoc.cs:
```diff
@@ -0,0 +1,25 @@
+
+namespace Libplanet.Explorer.Store
+{
+    public class AddressRefDoc
```
Why is this public while `TxRefDoc` is internal?

moreal replied:
It seems mistake. It should be also internal.

moreal committed:
Make RichStore implement IRichStore

RichStore became to implement IRichStore and to receive an instance of class
implemented `IStore` as the first parameter. So, it can wrap ㅕㅔ any store
implemented `IStore` now.

moreal committed:
Apply suggestions

Co-Authored-By: Hong Minhee <hong.minhee@gmail.com>

dahlia commented on Libplanet.Explorer.Executable/Program.cs:
```diff
@@ -40,11 +41,33 @@ public static async Task Main(string[] args)
+                    path: options.StorePath,
+                    flush: false,
+                    readOnly: options.Seeds is null
+                );
```
This seems equivalent to the first statement in the above clause.  Could you merge these two statements into a single statement and extract it from the `if`–`else` clauses?

dahlia commented on Libplanet.Explorer/Controllers/ExplorerController.cs:
```diff
@@ -37,8 +39,9 @@ public IActionResult GetRoot()
             var json = _schema.Execute(_ =>
             {
-                _.UserContext = BlockChain;
+                _.UserContext = (BlockChain, Store);
```
Rather than an ad-hoc tuple, how about defining a new type having two fields, say, `ExplorerContext` instead?

moreal replied:
Okay. To achieve it, we can use `IBlockChainContext<T>` type directly instead of defining a new type.
